# Data Cleansing and Feature Engineering Activity
Use this notebook to perform some data cleansing and feature engineering on the London Fire Brigade data that can be found in 

datasets/London Fire Brigade/london_fire_brigade.csv

The original data came from here:

https://www.kaggle.com/jboysen/london-fire

## Import libraries

In [1]:
from dasi_library import *

## Load data

In [2]:
df = readCsv("../datasets/London Fire Brigade/london_fire_brigade.csv")
df

borough_name timestamp_of_call postcode_full  easting_rounded  \
0                   CAMDEN  20/01/2017 08:57       W1T 7PD           529450   
1                   NEWHAM  21/04/2017 17:42           NaN           539650   
2           WALTHAM FOREST  20/01/2017 18:21       E17 6RH           536950   
3                   EALING  07/03/2017 11:27       W13 9YN           516650   
4               WANDSWORTH  04/02/2017 17:31      SW15 2SP           524250   
5            TOWER HAMLETS  08/01/2017 16:54           NaN           537350   
6                  BROMLEY  17/03/2017 13:29           NaN           547750   
7                  LAMBETH  26/01/2017 17:25       SW4 9DT           529050   
8                  ENFIELD  13/01/2017 15:42           NaN           537450   
9               HILLINGDON  08/03/2017 10:30       UB8 3QP           506850   
10                  BEXLEY  19/02/2017 00:15      DA14 6EN           546450   
11                 ENFIELD  02/04/2017 12:03           NaN           534050   
12                 HACKNEY  23/02/2017 14:36           NaN           532750   
13                HOUNSLOW  23/04/2017 21:27           NaN           514450   
14                HOUNSLOW  27/04/2017 08:32        W4 3BD           519450   
15           NOT GEO-CODED  01/01/2017 15:30           NaN           541450   
16  HAMMERSMITH AND FULHAM  08/02/2017 18:34       SW6 6PQ           523450   
17                  BEXLEY  25/02/2017 15:06           NaN           545750   
18                  MERTON  15/04/2017 07:10      SW19 1RF           524950   
19                 BROMLEY  26/03/2017 18:30       BR5 4HQ           547550   
20                  MERTON  26/01/2017 19:57           NaN           527950   
21                 ENFIELD  08/03/2017 08:35        N9 8NZ           534950   
22                 BROMLEY  12/03/2017 13:26           NaN           534850   
23                  CAMDEN  29/04/2017 04:09           NaN           529150   
24                 LAMBETH  03/01/2017 09:15           NaN           529650   
25                 LAMBETH  29/03/2017 11:13      SW16 1QL           529550   
26    KINGSTON UPON THAMES  20/01/2017 14:47       KT6 4JZ           517950   
27               SOUTHWARK  03/03/2017 18:55       SE5 7HN           532650   
28                 LAMBETH  22/01/2017 00:33           NaN           531350   
29                  NEWHAM  20/01/2017 21:29           NaN           541050   
30               ISLINGTON  28/03/2017 01:40        N1 0EP           531150   
31                 CROYDON  03/01/2017 08:29           NaN           532650   
32                  CAMDEN  25/01/2017 23:38           NaN           527150   
33                 LAMBETH  25/04/2017 23:35           NaN           529950   
34                 LAMBETH  01/02/2017 06:24           NaN           531350   
35           TOWER HAMLETS  04/03/2017 11:25           NaN           534050   
36                  NEWHAM  22/01/2017 21:42           NaN           542450   
37                 CROYDON  09/04/2017 15:36           NaN           532550   
38  KENSINGTON AND CHELSEA  19/02/2017 08:40           NaN           526150   
39               SOUTHWARK  30/01/2017 14:19           NaN           532850   
40    KINGSTON UPON THAMES  31/01/2017 00:56           NaN           517850   
41                  BARNET  05/01/2017 17:23           NaN           526050   
42               SOUTHWARK  13/02/2017 11:31           NaN           533650   
43           TOWER HAMLETS  27/04/2017 12:20        E2 7RU           533950   
44                  BARNET  18/02/2017 21:38           NaN           525950   
45                 ENFIELD  18/02/2017 18:57       N14 4NL           529250   
46                 ENFIELD  15/01/2017 10:42       EN1 1YP           534150   
47               REDBRIDGE  07/03/2017 20:33       IG6 3TN           546550   
48                  BARNET  15/02/2017 16:43       N20 9LZ           526250   
49              HILLINGDON  08/02/2017 17:17           NaN      

## Check data types

In [3]:
df.dtypes

borough_name                                  object
timestamp_of_call                             object
postcode_full                                 object
easting_rounded                                int64
northing_rounded                               int64
property_category                             object
property_type                                 object
incident_group                                object
first_pump_arriving_attendance_time          float64
first_pump_arriving_deployed_from_station     object
num_pumps_attending                          float64
dtype: object

## Check for nulls

In [4]:
checkForNulls(df)

borough_name                                 0.00
timestamp_of_call                            0.00
postcode_full                                0.58
easting_rounded                              0.00
northing_rounded                             0.00
property_category                            0.00
property_type                                0.00
incident_group                               0.00
first_pump_arriving_attendance_time          0.06
first_pump_arriving_deployed_from_station    0.06
num_pumps_attending                          0.08
dtype: float64

## Data cleansing
Think about what data cleansing we need to do

In [5]:
# TODO

The timestamp_of_call feature has a string data type, but should be datetime:

In [6]:
df["timestamp_of_call"] = convertToDateTime(df, "timestamp_of_call")

The num_pumps_attending feature has some nulls.  We can impute with the median:

In [7]:
df = imputeNullWithMedian(df, "num_pumps_attending")

Check the unique list of borough names:

In [8]:
listUnique(df, "borough_name")

['CAMDEN',
 'NEWHAM',
 'WALTHAM FOREST',
 'EALING',
 'WANDSWORTH',
 'TOWER HAMLETS',
 'BROMLEY',
 'LAMBETH',
 'ENFIELD',
 'HILLINGDON',
 'BEXLEY',
 'HACKNEY',
 'HOUNSLOW',
 'NOT GEO-CODED',
 'HAMMERSMITH AND FULHAM',
 'MERTON',
 'KINGSTON UPON THAMES',
 'SOUTHWARK',
 'ISLINGTON',
 'CROYDON',
 'KENSINGTON AND CHELSEA',
 'BARNET',
 'REDBRIDGE']

There is one value 'NOT GEO-CODED' that is effectively a null.  We could replace with null:

In [9]:
df = replaceValues(df, "borough_name", 'NOT GEO-CODED', np.nan)
df

borough_name   timestamp_of_call postcode_full  easting_rounded  \
0                   CAMDEN 2017-01-20 08:57:00       W1T 7PD           529450   
1                   NEWHAM 2017-04-21 17:42:00           NaN           539650   
2           WALTHAM FOREST 2017-01-20 18:21:00       E17 6RH           536950   
3                   EALING 2017-07-03 11:27:00       W13 9YN           516650   
4               WANDSWORTH 2017-04-02 17:31:00      SW15 2SP           524250   
5            TOWER HAMLETS 2017-08-01 16:54:00           NaN           537350   
6                  BROMLEY 2017-03-17 13:29:00           NaN           547750   
7                  LAMBETH 2017-01-26 17:25:00       SW4 9DT           529050   
8                  ENFIELD 2017-01-13 15:42:00           NaN           537450   
9               HILLINGDON 2017-08-03 10:30:00       UB8 3QP           506850   
10                  BEXLEY 2017-02-19 00:15:00      DA14 6EN           546450   
11                 ENFIELD 2017-02-04 12:03:00           NaN           534050   
12                 HACKNEY 2017-02-23 14:36:00           NaN           532750   
13                HOUNSLOW 2017-04-23 21:27:00           NaN           514450   
14                HOUNSLOW 2017-04-27 08:32:00        W4 3BD           519450   
15                     NaN 2017-01-01 15:30:00           NaN           541450   
16  HAMMERSMITH AND FULHAM 2017-08-02 18:34:00       SW6 6PQ           523450   
17                  BEXLEY 2017-02-25 15:06:00           NaN           545750   
18                  MERTON 2017-04-15 07:10:00      SW19 1RF           524950   
19                 BROMLEY 2017-03-26 18:30:00       BR5 4HQ           547550   
20                  MERTON 2017-01-26 19:57:00           NaN           527950   
21                 ENFIELD 2017-08-03 08:35:00        N9 8NZ           534950   
22                 BROMLEY 2017-12-03 13:26:00           NaN           534850   
23                  CAMDEN 2017-04-29 04:09:00           NaN           529150   
24                 LAMBETH 2017-03-01 09:15:00           NaN           529650   
25                 LAMBETH 2017-03-29 11:13:00      SW16 1QL           529550   
26    KINGSTON UPON THAMES 2017-01-20 14:47:00       KT6 4JZ           517950   
27               SOUTHWARK 2017-03-03 18:55:00       SE5 7HN           532650   
28                 LAMBETH 2017-01-22 00:33:00           NaN           531350   
29                  NEWHAM 2017-01-20 21:29:00           NaN           541050   
30               ISLINGTON 2017-03-28 01:40:00        N1 0EP           531150   
31                 CROYDON 2017-03-01 08:29:00           NaN           532650   
32                  CAMDEN 2017-01-25 23:38:00           NaN           527150   
33                 LAMBETH 2017-04-25 23:35:00           NaN           529950   
34                 LAMBETH 2017-01-02 06:24:00           NaN           531350   
35           TOWER HAMLETS 2017-04-03 11:25:00           NaN           534050   
36                  NEWHAM 2017-01-22 21:42:00           NaN           542450   
37                 CROYDON 2017-09-04 15:36:00           NaN           532550   
38  KENSINGTON AND CHELSEA 2017-02-19 08:40:00           NaN           526150   
39               SOUTHWARK 2017-01-30 14:19:00           NaN           532850   
40    KINGSTON UPON THAMES 2017-01-31 00:56:00           NaN           517850   
41                  BARNET 2017-05-01 17:23:00           NaN           526050   
42               SOUTHWARK 2017-02-13 11:31:00           NaN           533650   
43           TOWER HAMLETS 2017-04-27 12:20:00        E2 7RU           533950   
44                  BARNET 2017-02-18 21:38:00           NaN           525950   
45                 ENFIELD 2017-02-18 18:57:00       N14 4NL           529250   
46                 ENFIELD 2017-01-15 10:42:00       EN1 1YP           534150   
47               REDBRIDGE 2017-07-03 20:33:00       IG6 3TN           546550   
48                  BARNET 2017-02-15 16:43:0

But we still need to deal with the fact that it is null.  Looking at the data, we see we have eastings and northings.  We can use this to look up the borough.  We could use something like: https://gridreferencefinder.com/, or there are software packages that will convert geographic coordinates to location names.  Here is [another resource](https://www.google.com/maps/d/u/0/viewer?ptab=2&ie=UTF8&oe=UTF8&msa=0&mid=1t4G7Q0brBWa2_kKwYyEtoxmCd60&ll=51.588246412761684%2C-0.12131868990172734&z=11) that shows the borough boundaries.

Looking up E 541450 N 194050, we see the coordinate is in Essex, just outside London, so we need to decide how to encode this.  Let's replace with "ESSEX":

In [10]:
setValue(df, 15, "borough_name", "ESSEX")

In [11]:
df

borough_name   timestamp_of_call postcode_full  easting_rounded  \
0                   CAMDEN 2017-01-20 08:57:00       W1T 7PD           529450   
1                   NEWHAM 2017-04-21 17:42:00           NaN           539650   
2           WALTHAM FOREST 2017-01-20 18:21:00       E17 6RH           536950   
3                   EALING 2017-07-03 11:27:00       W13 9YN           516650   
4               WANDSWORTH 2017-04-02 17:31:00      SW15 2SP           524250   
5            TOWER HAMLETS 2017-08-01 16:54:00           NaN           537350   
6                  BROMLEY 2017-03-17 13:29:00           NaN           547750   
7                  LAMBETH 2017-01-26 17:25:00       SW4 9DT           529050   
8                  ENFIELD 2017-01-13 15:42:00           NaN           537450   
9               HILLINGDON 2017-08-03 10:30:00       UB8 3QP           506850   
10                  BEXLEY 2017-02-19 00:15:00      DA14 6EN           546450   
11                 ENFIELD 2017-02-04 12:03:00           NaN           534050   
12                 HACKNEY 2017-02-23 14:36:00           NaN           532750   
13                HOUNSLOW 2017-04-23 21:27:00           NaN           514450   
14                HOUNSLOW 2017-04-27 08:32:00        W4 3BD           519450   
15                   ESSEX 2017-01-01 15:30:00           NaN           541450   
16  HAMMERSMITH AND FULHAM 2017-08-02 18:34:00       SW6 6PQ           523450   
17                  BEXLEY 2017-02-25 15:06:00           NaN           545750   
18                  MERTON 2017-04-15 07:10:00      SW19 1RF           524950   
19                 BROMLEY 2017-03-26 18:30:00       BR5 4HQ           547550   
20                  MERTON 2017-01-26 19:57:00           NaN           527950   
21                 ENFIELD 2017-08-03 08:35:00        N9 8NZ           534950   
22                 BROMLEY 2017-12-03 13:26:00           NaN           534850   
23                  CAMDEN 2017-04-29 04:09:00           NaN           529150   
24                 LAMBETH 2017-03-01 09:15:00           NaN           529650   
25                 LAMBETH 2017-03-29 11:13:00      SW16 1QL           529550   
26    KINGSTON UPON THAMES 2017-01-20 14:47:00       KT6 4JZ           517950   
27               SOUTHWARK 2017-03-03 18:55:00       SE5 7HN           532650   
28                 LAMBETH 2017-01-22 00:33:00           NaN           531350   
29                  NEWHAM 2017-01-20 21:29:00           NaN           541050   
30               ISLINGTON 2017-03-28 01:40:00        N1 0EP           531150   
31                 CROYDON 2017-03-01 08:29:00           NaN           532650   
32                  CAMDEN 2017-01-25 23:38:00           NaN           527150   
33                 LAMBETH 2017-04-25 23:35:00           NaN           529950   
34                 LAMBETH 2017-01-02 06:24:00           NaN           531350   
35           TOWER HAMLETS 2017-04-03 11:25:00           NaN           534050   
36                  NEWHAM 2017-01-22 21:42:00           NaN           542450   
37                 CROYDON 2017-09-04 15:36:00           NaN           532550   
38  KENSINGTON AND CHELSEA 2017-02-19 08:40:00           NaN           526150   
39               SOUTHWARK 2017-01-30 14:19:00           NaN           532850   
40    KINGSTON UPON THAMES 2017-01-31 00:56:00           NaN           517850   
41                  BARNET 2017-05-01 17:23:00           NaN           526050   
42               SOUTHWARK 2017-02-13 11:31:00           NaN           533650   
43           TOWER HAMLETS 2017-04-27 12:20:00        E2 7RU           533950   
44                  BARNET 2017-02-18 21:38:00           NaN           525950   
45                 ENFIELD 2017-02-18 18:57:00       N14 4NL           529250   
46                 ENFIELD 2017-01-15 10:42:00       EN1 1YP           534150   
47               REDBRIDGE 2017-07-03 20:33:00       IG6 3TN           546550   
48                  BARNET 2017-02-15 16:43:0

We also have an issue with missing postcodes.  We can use the same approach, but with so many missing postcodes we would want to use an automated tool to find a postcode based on the easting and northing.  I won't do that now, but if I was doing this for real I would look into this further.

We have nulls in the attendance time.  Let's impute with the median:

In [12]:
df = imputeNullWithMedian(df, "first_pump_arriving_attendance_time")

We also have nulls in the first_pump_arriving_deployed_from_station column.  It would be hard to impute that based on other column data.  Our best bet is to look for the station closest to the incident (we would need to source a list of fire stations).  As I don't have that handy, I will impute with "UNKNOWN":


In [13]:
df = imputeNullWithValue(df, "first_pump_arriving_deployed_from_station", "UNKNOWN")

Let's check our situation:

In [14]:
checkForNulls(df)

borough_name                                 0.00
timestamp_of_call                            0.00
postcode_full                                0.58
easting_rounded                              0.00
northing_rounded                             0.00
property_category                            0.00
property_type                                0.00
incident_group                               0.00
first_pump_arriving_attendance_time          0.00
first_pump_arriving_deployed_from_station    0.00
num_pumps_attending                          0.00
dtype: float64

So apart from postcode (which we discussed above) everything else has had nulls removed.

## Feature engineering
Think about what feature engineering we need to do

In [15]:
# TODO

Let's split out the postcode (bearing in mind that we haven't yet replaced all the null postcodes):

In [16]:
df = splitFeatureOnSeparator(df, "postcode_full", " ", ["postcode1", "postcode2"])

In [17]:
df

borough_name   timestamp_of_call  easting_rounded  \
0                   CAMDEN 2017-01-20 08:57:00           529450   
1                   NEWHAM 2017-04-21 17:42:00           539650   
2           WALTHAM FOREST 2017-01-20 18:21:00           536950   
3                   EALING 2017-07-03 11:27:00           516650   
4               WANDSWORTH 2017-04-02 17:31:00           524250   
5            TOWER HAMLETS 2017-08-01 16:54:00           537350   
6                  BROMLEY 2017-03-17 13:29:00           547750   
7                  LAMBETH 2017-01-26 17:25:00           529050   
8                  ENFIELD 2017-01-13 15:42:00           537450   
9               HILLINGDON 2017-08-03 10:30:00           506850   
10                  BEXLEY 2017-02-19 00:15:00           546450   
11                 ENFIELD 2017-02-04 12:03:00           534050   
12                 HACKNEY 2017-02-23 14:36:00           532750   
13                HOUNSLOW 2017-04-23 21:27:00           514450   
14                HOUNSLOW 2017-04-27 08:32:00           519450   
15                   ESSEX 2017-01-01 15:30:00           541450   
16  HAMMERSMITH AND FULHAM 2017-08-02 18:34:00           523450   
17                  BEXLEY 2017-02-25 15:06:00           545750   
18                  MERTON 2017-04-15 07:10:00           524950   
19                 BROMLEY 2017-03-26 18:30:00           547550   
20                  MERTON 2017-01-26 19:57:00           527950   
21                 ENFIELD 2017-08-03 08:35:00           534950   
22                 BROMLEY 2017-12-03 13:26:00           534850   
23                  CAMDEN 2017-04-29 04:09:00           529150   
24                 LAMBETH 2017-03-01 09:15:00           529650   
25                 LAMBETH 2017-03-29 11:13:00           529550   
26    KINGSTON UPON THAMES 2017-01-20 14:47:00           517950   
27               SOUTHWARK 2017-03-03 18:55:00           532650   
28                 LAMBETH 2017-01-22 00:33:00           531350   
29                  NEWHAM 2017-01-20 21:29:00           541050   
30               ISLINGTON 2017-03-28 01:40:00           531150   
31                 CROYDON 2017-03-01 08:29:00           532650   
32                  CAMDEN 2017-01-25 23:38:00           527150   
33                 LAMBETH 2017-04-25 23:35:00           529950   
34                 LAMBETH 2017-01-02 06:24:00           531350   
35           TOWER HAMLETS 2017-04-03 11:25:00           534050   
36                  NEWHAM 2017-01-22 21:42:00           542450   
37                 CROYDON 2017-09-04 15:36:00           532550   
38  KENSINGTON AND CHELSEA 2017-02-19 08:40:00           526150   
39               SOUTHWARK 2017-01-30 14:19:00           532850   
40    KINGSTON UPON THAMES 2017-01-31 00:56:00           517850   
41                  BARNET 2017-05-01 17:23:00           526050   
42               SOUTHWARK 2017-02-13 11:31:00           533650   
43           TOWER HAMLETS 2017-04-27 12:20:00           533950   
44                  BARNET 2017-02-18 21:38:00           525950   
45                 ENFIELD 2017-02-18 18:57:00           529250   
46                 ENFIELD 2017-01-15 10:42:00           534150   
47               REDBRIDGE 2017-07-03 20:33:00           546550   
48                  BARNET 2017-02-15 16:43:00           526250   
49              HILLINGDON 2017-08-02 17:17:00           507750   

    northing_rounded  property_category  \
0             182050    Non Residential   
1             183750           Dwelling   
2             189350    Non Residential   
3             180250    Non Residential   
4             174950    Non Residential   
5             181850           Dwelling   
6             168450           Dwelling   
7             174650  Other Residential   
8             198050           Dwelling   
9             182850  Other Residential   
10            171650  Other Residential   
11            192050           Dwelling   
12            183650           Dwelli

Let's also split out the call date and time:

In [18]:
df = splitFeatureDate(df, "timestamp_of_call")

In [19]:
df

borough_name  easting_rounded  northing_rounded  \
0                   CAMDEN           529450            182050   
1                   NEWHAM           539650            183750   
2           WALTHAM FOREST           536950            189350   
3                   EALING           516650            180250   
4               WANDSWORTH           524250            174950   
5            TOWER HAMLETS           537350            181850   
6                  BROMLEY           547750            168450   
7                  LAMBETH           529050            174650   
8                  ENFIELD           537450            198050   
9               HILLINGDON           506850            182850   
10                  BEXLEY           546450            171650   
11                 ENFIELD           534050            192050   
12                 HACKNEY           532750            183650   
13                HOUNSLOW           514450            175750   
14                HOUNSLOW           519450            178050   
15                   ESSEX           541450            194050   
16  HAMMERSMITH AND FULHAM           523450            177050   
17                  BEXLEY           545750            175250   
18                  MERTON           524950            170550   
19                 BROMLEY           547550            166950   
20                  MERTON           527950            170650   
21                 ENFIELD           534950            194550   
22                 BROMLEY           534850            169550   
23                  CAMDEN           529150            183750   
24                 LAMBETH           529650            171050   
25                 LAMBETH           529550            171650   
26    KINGSTON UPON THAMES           517950            167250   
27               SOUTHWARK           532650            177050   
28                 LAMBETH           531350            175650   
29                  NEWHAM           541050            183950   
30               ISLINGTON           531150            183850   
31                 CROYDON           532650            166850   
32                  CAMDEN           527150            185150   
33                 LAMBETH           529950            171150   
34                 LAMBETH           531350            175650   
35           TOWER HAMLETS           534050            182450   
36                  NEWHAM           542450            185550   
37                 CROYDON           532550            165950   
38  KENSINGTON AND CHELSEA           526150            178050   
39               SOUTHWARK           532850            178550   
40    KINGSTON UPON THAMES           517850            169550   
41                  BARNET           526050            188350   
42               SOUTHWARK           533650            178750   
43           TOWER HAMLETS           533950            183050   
44                  BARNET           525950            188650   
45                 ENFIELD           529250            195350   
46                 ENFIELD           534150            196550   
47               REDBRIDGE           546550            192150   
48                  BARNET           526250            194350   
49              HILLINGDON           507750            181550   

    property_category                                      property_type  \
0     Non Residential                               Purpose built office   
1            Dwelling  Purpose Built Flats/Maisonettes - Up to 3 storeys   
2     Non Residential                            Local Government Office   
3     Non Residential                               Underground car park   
4     Non Residential                                   Pub/wine bar/bar   
5            Dwelling   Purpose Built Flats/Maisonettes - 4 to 9 storeys   
6            Dwelling                           House - single occupancy   
7   Other Residential                          Nursing/Care Home/Hospice   
8            D

We may also want to one-hot-encode any features we want to use to build our models:

In [20]:
df = oneHotEncode(df, ["incident_group"])

In [21]:
df = oneHotEncode(df, ["property_category"])

In [22]:
df = oneHotEncode(df, ["borough_name"])

In [23]:
df

easting_rounded  northing_rounded  \
0            529450            182050   
1            539650            183750   
2            536950            189350   
3            516650            180250   
4            524250            174950   
5            537350            181850   
6            547750            168450   
7            529050            174650   
8            537450            198050   
9            506850            182850   
10           546450            171650   
11           534050            192050   
12           532750            183650   
13           514450            175750   
14           519450            178050   
15           541450            194050   
16           523450            177050   
17           545750            175250   
18           524950            170550   
19           547550            166950   
20           527950            170650   
21           534950            194550   
22           534850            169550   
23           529150            183750   
24           529650            171050   
25           529550            171650   
26           517950            167250   
27           532650            177050   
28           531350            175650   
29           541050            183950   
30           531150            183850   
31           532650            166850   
32           527150            185150   
33           529950            171150   
34           531350            175650   
35           534050            182450   
36           542450            185550   
37           532550            165950   
38           526150            178050   
39           532850            178550   
40           517850            169550   
41           526050            188350   
42           533650            178750   
43           533950            183050   
44           525950            188650   
45           529250            195350   
46           534150            196550   
47           546550            192150   
48           526250            194350   
49           507750            181550   

                                        property_type  \
0                                Purpose built office   
1   Purpose Built Flats/Maisonettes - Up to 3 storeys   
2                             Local Government Office   
3                                Underground car park   
4                                    Pub/wine bar/bar   
5    Purpose Built Flats/Maisonettes - 4 to 9 storeys   
6                            House - single occupancy   
7                           Nursing/Care Home/Hospice   
8   Purpose Built Flats/Maisonettes - Up to 3 storeys   
9                                         Hotel/motel   
10                                        Hotel/motel   
11   Purpose Built Flats/Maisonettes - 4 to 9 storeys   
12   Purpose Built Flats/Maisonettes - 4 to 9 storeys   
13                   Self contained Sheltered Housing   
14                        Retirement/Old Persons Home   
15  Purpose Built Flats/Maisonettes - Up to 3 storeys   
16                        Retirement/Old Persons Home   
17                           House - single occupancy   
18                               Purpose built office   
19                                         Tree scrub   
20     Converted Flat/Maisonettes - 3 or more storeys   
21                        Retirement/Old Persons Home   
22                   Self contained Sheltered Housing   
23   Purpose Built Flats/Maisonettes - 4 to 9 storeys   
24        Converted Flat/Maisonette - Up to 2 storeys   
25                                                Car   
26                     Small refuse/rubbish container   
27                                          Warehouse   
28  Purpose Built Flats/Maisonettes - 10 or more s...   
29                   Self contained Sheltered Housing   
30                  Hostel (e.g. for homeless people)   
31   Purpose Built Flats/Maisonettes - 4 to 9 storeys   
32     Converted Flat/Maisonett

There are other columns we could one-hot-encode if we decide to use them for our model.